File: course/string/H.cpp
#include <bits/stdc++.h>

using namespace std;

using ll = int64_t;
using ld = long double;

const char SPACE = ' ', ENDL = '\n';

//////////////////////////////////////////////////////////////////

template <typename T>
T read() {
    T value;
    cin >> value;
    return value;
}

int ri() {
    return read<int>();
}

ll rll() {
    return read<ll>();
}

string rs() {
    return read<string>();
}

//////////////////////////////////////////////////////////////////

bool yn(bool result, string yes = "YES", string no = "NO") {
	cout << (result ? yes : no) << ENDL;
	return result;
}

//////////////////////////////////////////////////////////////////

ld sqrt_ld(ld value) {
    return sqrt(value);
}

//////////////////////////////////////////////////////////////////

void init_io() {
    string file_name = "";
    if (!file_name.empty()) {
        string input_file_name = file_name + ".in";
        freopen(input_file_name.c_str(), "rt", stdin);

        string output_file_name = file_name + ".out";
        freopen(output_file_name.c_str(), "wt", stdout);
    } else {
        #ifndef ONLINE_JUDGE
        string input_file_name = "../../input.txt"; // reading from file in the same directory
        freopen(input_file_name.c_str(), "rt", stdin);
        #endif
    }

	ios::sync_with_stdio(false);
	cin.tie(nullptr);

	cout << setprecision(10) << fixed;
}

string signs = "-;_";

string refine(string& s) {
    string refined = "";
    for (char& ch : s) {
        if (string::npos != signs.find(ch)) continue;

        refined += tolower(ch);
    }

    return refined;
}

vector<bool> get_ans(vector<string>& parts, vector<string>& queries) {
    for (string& part : parts) {
        part = refine(part);
    }

    string a = parts[0], b = parts[1], c = parts[2];

    vector<string> permutations = {
        a + b + c,
        a + c + b,
        b + a + c,
        b + c + a,
        c + a + b,
        c + b + a
    };

    int q = queries.size();
    vector<bool> answers(q, false);

    for (int i = 0; i < q; ++i) {
        string query = refine(queries[i]);

        bool answer = false;
        for (string& p : permutations) {
            answer |= (p == query);
        }

        answers[i] = answer;
    }

    return answers;
}

void solve() {
    int n = 3;
    vector<string> parts(n);
    for (string& part : parts) {
        part = rs();
    }

    int q = ri();
    vector<string> queries(q);
    for (string& query : queries) {
        query = rs();
    }

    vector<bool> answers = get_ans(parts, queries);
    for (bool answer : answers) {
        yn(answer, "ACC", "WA");
    }
}

int main() {
    init_io();
    solve();
    return 0;
}

File: course/string/H.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.StringTokenizer;

public class H {

    private void solve() {
        int n = 3;
        String[] parts = new String[n];
        for (int i = 0; i < parts.length; ++i) {
            parts[i] = readString();
        }

        int q = readInt();
        String[] queries = new String[q];
        for (int i = 0; i < q; ++i) {
            queries[i] = readString();
        }

        boolean[] answers = getAnswers(parts, queries);
        for (boolean answer : answers) {
            yesNo(answer, "ACC", "WA");
        }
    }

    final String SIGNS = "-;_";

    String refine(String s) {
        for (char sign : SIGNS.toCharArray()) {
            s = s.replace("" + sign, "");
        }

        s = s.toLowerCase();

        return s;
    }

    boolean[] getAnswers(String[] parts, String[] queries) {
        int n = parts.length;
        for (int i = 0; i < n; ++i) {
            parts[i] = refine(parts[i]);
        }

        String a = parts[0], b = parts[1], c = parts[2];

        String[] permutations = {
                a + b + c,
                a + c + b,
                b + a + c,
                b + c + a,
                c + a + b,
                c + b + a
        };

        boolean[] answers = new boolean[queries.length];
        for (int i = 0; i < queries.length; ++i) {
            String query = refine(queries[i]);

            for (String permutation : permutations) {
                answers[i] |= permutation.equals(query);
            }
        }

        return answers;
    }

    //////////////////////////////////////////////////////////////////

    long sqrtLong(long x) {
        long root = (long)Math.sqrt(x);
        while (root * root > x) --root;
        while ((root + 1) * (root + 1) <= x) ++root;
        return root;
    }

    //////////////////////////////////////////////////////////////////

    private boolean yesNo(boolean yes) {
        return yesNo(yes, "YES", "NO");
    }

    private boolean yesNo(boolean yes, String yesString, String noString) {
        out.println(yes ? yesString : noString);
        return yes;
    }

    //////////////////////////////////////////////////////////////////

    private long readLong() {
        return Long.parseLong(readString());
    }

    private int[] readIntArray(int n) {
        int[] a = new int[n];
        for (int i = 0; i < n; ++i) a[i] = readInt();
        return a;
    }

    private int readInt() {
        return Integer.parseInt(readString());
    }

    private String readString() {
        while (!tok.hasMoreTokens()) {
            String nextLine = readLine();
            if (null == nextLine) return null;
            tok = new StringTokenizer(nextLine);
        }

        return tok.nextToken();
    }

    private String readLine() {
        try {
            return in.readLine();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    //////////////////////////////////////////////////////////////////

    private BufferedReader in;
    private StringTokenizer tok;
    private PrintWriter out;

    private void initFileIO(String inputFileName, String outputFileName) throws FileNotFoundException {
        in = new BufferedReader(new FileReader(inputFileName));
        out = new PrintWriter(outputFileName);
    }

    private void initConsoleIO() throws IOException {
        in = new BufferedReader(new InputStreamReader(System.in));
        out = new PrintWriter(System.out);
    }

    private void initIO() throws IOException {
        Locale.setDefault(Locale.US);

        String fileName = "";

        if (!fileName.isEmpty()) {
            initFileIO(fileName + ".in", fileName + ".out");
        } else {
            if (new File("input.txt").exists()) {
                initFileIO("input.txt", "output.txt");
            } else {
                initConsoleIO();
            }
        }

        tok = new StringTokenizer("");
    }

    //////////////////////////////////////////////////////////////////

    private void run() {
        try {
            long timeStart = System.currentTimeMillis();

            initIO();
            solve();
            out.close();

            long timeEnd = System.currentTimeMillis();
            System.err.println("Time(ms) = " + (timeEnd - timeStart));
        } catch (Exception e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }

    public static void main(String[] args) {
        new H().run();
    }
}
